### PR TITLE
Decrease total sales per product when an order is reversed

### LINKED ISSUE #23796: Decrease total sales per product when an order is reversed
**Is your feature request related to a problem? Please describe.**
The `total_sales` value which is stored per product (in the meta data, as well as the lookup table), is an increment only value. Currently it's only incremented when the order status is changed to either processing, completed or refunded.

However this value is not decreased/adjusted when for example one of the following scenarios happens:

- Order is refunded / cancelled
- A test order is made and then deleted
- An order is adjusted and the quantity of a particular line item is changed

Over time this can lead to a large discrepancy with the amount of times a product has actually been sold. For example if we compare it with the WooCommerce Report > Sales by product > Top sellers. This report calculates the totals within the orders instead of looking at the `total_sales` value. 

**Describe the solution you'd like**
Ideally I'd expect the `total_sales` value to be update as soon as any information in the order changes or the status of the order changes which negates it from being an order with sold products.
This would be similar to the scenario where stock is automatically increased in certain scenarios, similarly the `total_sales` value would be readjusted.

**Describe alternatives you've considered**
A less then ideal solution would to create a tool which would allow re-syncing the values by looking at the actual orders. This would be similar to the count data which shows up in WooCommerce Report > Sales by product > Top sellers, but then having these values saved per product in the `total_sales` value.

**Additional context**
Currently the `total_sales` value for each product is being used in the following locations:

- Shortcode for displaying product with `best_selling="1"`
- Sort by popularity on the product pages
- Product widget when order by sales is selected

### PR DESCRIPTION
**Pull request by jonathansadowski**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

* Add new functions to decrement `total_sales` in an order for canceled, refunded, and deleted orders.
* Adds post meta to persist that we've recorded cancelation (`_recorded_cancelations`).
* Updates the method that increments `total_sales` to clear `_recorded_cancelations`, so that toggling an order between statuses will properly update the total.

Closes #23796 .

### How to test the changes in this Pull Request:

1. Create a new product, and observe its current `total_sales` value.
2. Add that product to your cart, and checkout.
3. Verify that the `total_sales` value was increased by the quantity of that product in your cart.
4. Change the status of the order to `canceled`. Verify that the `total_sales` value was decreased by the quantity of that product in the order.
5. Change the status of the order to `refunded`. Verify that the `total_sales` value of the product remains unchanged.
6. Change the status of the order to `processing`. Verify that the `total_sales` value of the product is increased by the quantity of that product in the order.
7. Trash and then permanently delete the order. Verify that the `total_sales` value of the product is decreased by the quantity of that product in the order.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

> Refunds, cancelations, and deletions of an order will now cause `total_sales` for a product to be decremented by the quantity of the product in the order.

### CODE REVIEW

**Comment by peterfabian** [comment]
Hi, sorry it took so long to get to this. 

Nice work here. 

Besides moving orders from one status to another, there is also a possibility of a refund (without cancelling or deleting the order), and it seems to me we should probably take it into account in this solution. 
Screenshot 2020-05-06 at 14 37 02

What I've tried:
- Trashing the order does not change the total_sales. Not sure if it should, that status is a bit ambiguous.
- Deleting order decreases the total_sales correctly ✅
- Cancelling an order decreases the total_sales correctly ✅
- Putting the order back into Processing status increases the total_sales correctly back ✅
- Changing the order to Refunded decreases the total_sales correctly ✅
- Putting the order back into Processing status increases the total_sales back, not 100% sure if that is correct, because the order now looks as if all the items got refunded:
Screenshot 2020-05-06 at 14 45 24
This is a bit of an edge case scenario I think, but perhaps the total_sales should not go back up until the refund is actually _removed_. 

Maybe I'm wrong, but it seems to me the `_recorded_cancelations` is somewhat superfluous and we should be able to get this number when we sum up sales on the order + refunds related to the order? We could still utilize `_recorded_cancelations` as a cache, or perhaps there's a use case  I'm not seeing where it is required, but I'm somewhat cautious about adding new meta fields unless needed. Please let me know what you think, thanks!

**Comment by jonathansadowski** [comment]
> Maybe I'm wrong, but it seems to me the _recorded_cancelations is somewhat superfluous and we should be able to get this number when we sum up sales on the order + refunds related to the order? We could still utilize _recorded_cancelations as a cache, or perhaps there's a use case I'm not seeing where it is required, but I'm somewhat cautious about adding new meta fields unless needed. Please let me know what you think, thanks!

Thanks for the feedback, I was able to rework the PR to remove that and introduce some more tests to verify trashed orders get update appropriately. Because trashes orders in the reports aren't counted as sales, I followed the same pattern.

File: includes/data-stores/class-wc-product-data-store-cpt.php [hunk 1]
```diff
@@ -1414,6 +1414,39 @@ public function update_product_stock( $product_id_with_stock, $stock_quantity =
...
+			"INNER JOIN {$wpdb->prefix}woocommerce_order_itemmeta AS order_item_meta__product_id_array ON " .
+			'order_items.order_item_id = order_item_meta__product_id_array.order_item_id ' .
+			"WHERE {$exclude_order_id_clause} posts.post_type IN ( 'shop_order','shop_order_refund' ) " .
+			"AND posts.post_status IN ( 'wc-completed', 'wc-processing', 'wc-on-hold' ) " .
+			"AND ( order_item_meta__product_id_array.meta_key IN ( '_product_id', '_variation_id' ) " .
+			'AND order_item_meta__product_id_array.meta_value = %d )';
+		$prepare_args[] = $product_id;
+		// phpcs:ignore WordPress.DB.PreparedSQL.NotPrepared
+		$total_sales = $wpdb->get_var( $wpdb->prepare( $query, $prepare_args ) );
+		$this->update_product_sales( $product_id, absint( $total_sales ), 'set' );
```
Comment (rrennick) [inline includes/data-stores/class-wc-product-data-store-cpt.php:1447]: @jonathansadowski `update_product_sales()` stores floating point `total_sales`. `abs( floatval( $total_sales ) )` should generate the comparable FP value.

File: includes/data-stores/class-wc-product-data-store-cpt.php [hunk 2]
```diff
@@ -1414,6 +1414,39 @@ public function update_product_stock( $product_id_with_stock, $stock_quantity =
...
+		}
+
+		$query = "SELECT SUM( order_item_meta__qty.meta_value ) AS order_item_count FROM {$wpdb->posts} AS posts " .
+			"INNER JOIN {$wpdb->prefix}woocommerce_order_items AS order_items ON posts.ID = order_items.order_id " .
+			"INNER JOIN {$wpdb->prefix}woocommerce_order_itemmeta AS order_item_meta__qty ON " .
+			"order_items.order_item_id = order_item_meta__qty.order_item_id AND order_item_meta__qty.meta_key = '_qty' " .
+			"INNER JOIN {$wpdb->prefix}woocommerce_order_itemmeta AS order_item_meta__product_id_array ON " .
+			'order_items.order_item_id = order_item_meta__product_id_array.order_item_id ' .
+			"WHERE {$exclude_order_id_clause} posts.post_type IN ( 'shop_order','shop_order_refund' ) " .
+			"AND posts.post_status IN ( 'wc-completed', 'wc-processing', 'wc-on-hold' ) " .
```
Comment (rrennick) [inline includes/data-stores/class-wc-product-data-store-cpt.php:1441]: I refunded half of the the test orders product count. The updated `total_sales` was `0.0`. Does `wc-refunded` need to be included in this list to balance the negative amount in the refund order?

**Comment by peterfabian** [comment]
I'm unassigning from 4.3 since there is some feedback that needs to be addressed.

File: includes/data-stores/class-wc-product-data-store-cpt.php (hunk 2 above)
Comment (jonathansadowski) [reply includes/data-stores/class-wc-product-data-store-cpt.php:1441]: Good catch. Thanks. I updated the query and added a new test to handle refunds.

**Comment by jonathansadowski** [comment]
Updated to account for refunds, and updated the tests. I'm having issues with the new test passing unless I run it in isolation. Digging into that...

**Pull request closed without merging**